src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorOptimizedFn.java

```diff
@@ -164,411 +138,16 @@ private void init(ProcessContext c) {
-        final byte[] baq = new byte[read.getLength()];
-        Arrays.fill(baq, NO_BAQ_UNCERTAINTY);
-        return baq;
-    }
-
-    /**
-     * Compute an actual BAQ array for read, based on its quals and the reference sequence
-     *
-     * @param read the read to BAQ
-     * @return a non-null BAQ tag array for read
-     */
-    private byte[] calculateBAQArray(final GATKRead read, ReferenceDataSource refDS) {
-        baq.baqRead(read, refDS, BAQ.CalculationMode.RECALCULATE, BAQ.QualityMode.ADD_TAG);
-        return BAQ.getBAQTag(read);
-    }
 
-    private ReadTransformer makeReadTransform() {
-        ReadTransformer f0 = BaseRecalibratorOptimizedFn::consolidateCigar;
-
-        ReadTransformer f =
-            f0.andThen(this::setDefaultBaseQualities)
-                .andThen(this::resetOriginalBaseQualities)
-                .andThen(ReadClipper::hardClipAdaptorSequence)
-                .andThen(ReadClipper::hardClipSoftClippedBases);
-
-        return f;
-    }
-
-    private static GATKRead consolidateCigar(GATKRead read) {
-        // Always consolidate the cigar string into canonical form, collapsing zero-length / repeated cigar elements.
-        // Downstream code cannot necessarily handle non-consolidated cigar strings.
-        read.setCigar(AlignmentUtils.consolidateCigar(read.getCigar()));
-        return read;
-    }
-
-    private GATKRead setDefaultBaseQualities(GATKRead read) {
-        // if we are using default quals, check if we need them, and add if necessary.
-        // 1. we need if reads are lacking or have incomplete quality scores
-        // 2. we add if defaultBaseQualities has a positive value
-        if (BRAC.defaultBaseQualities < 0) {
-            return read;
-        }
-        byte reads[] = read.getBases();
-        byte quals[] = read.getBaseQualities();
-        if (quals == null || quals.length < reads.length) {
-            byte new_quals[] = new byte[reads.length];
-            Arrays.fill(new_quals, BRAC.defaultBaseQualities);
-            read.setBaseQualities(new_quals);
-        }
-        return read;
-    }
-
-    private GATKRead resetOriginalBaseQualities(GATKRead read) {
-        if (!BRAC.useOriginalBaseQualities) {
-            return read;
-        }
-        return ReadUtils.resetOriginalBaseQualities(read);
+        recalibrationEngine = new BaseRecalibrationEngine(recalArgs, header);
     }
 
     private void saveReport(File output, RecalibrationTables rt, QuantizationInfo quantizationInfo, StandardCovariateList finalRequestedCovariates) throws FileNotFoundException {
-        // luckily we don't use BRAC.RAC.RECAL_TABLE without first coming through here, so this is fine.
-        BRAC.RAC.RECAL_TABLE = new PrintStream(output);
-        RecalUtils.outputRecalibrationReport(BRAC.RAC, quantizationInfo, rt, finalRequestedCovariates, BRAC.RAC.SORT_BY_ALL_COLUMNS);
+        try ( PrintStream reportStream = new PrintStream(output) ) {
```

Comment:
You know, if saveReport now deals with output streams, we should make that the interface (instead of a file) so that we can write to GCS again. Like this:

```java
saveReport(OutputStream output, ...) {
try (PrintStream reportStream = new PrintStream(output) { 
```

and we'd call it like this: `BaseRecalibratorDataflow.saveReport(BucketUtils.createFile(output, pipelineOptions), ...`

Alternatively we can pass the url (String) to saveReport, and it can call createFile internally.